### PR TITLE
FEA Deduplication

### PR DESCRIPTION
**Pull request by mjboos**
This PR implements a basic `deduplicate` function (see discussion in #306 ).
It is based on computing n-gram tfidf distances between unique words in a column (same as for `fuzzy_join`), hierarchically clustering the distance matrix, then deduplicating by picking the most frequent exemplar in each cluster as the "correct" spelling.

Open design question: Should this be a function or a class? It clearly has a state (and could be useful when "trained" on a large set of dirty categories and then applied to new incoming data) - but saving a translation table in a class would be costly. An alternative would be to go from a clustered distance matrix to a K-means representation based on string distances (where each centroid is the most frequent cluster exemplar).

It's currently WIP because I still want to add an example/documentation.

FWIW this function was immediately useful in my work, but its usefulness depends on the structure of your data:
If there are few underlying categories, which might be corrupted by typos or alternative spellings, then it should work well out of the box.

### CODE REVIEW

**Comment by jovan-stojanovic** [comment]
Hi @mjboos, thanks for opening this PR, exciting work ahead.
Just so you know, on Wednesday during the day we plan to have a sprint during which we will take a closer look at the problem.
We may come up with some comments and answers to your questions.

**Comment by mjboos** [comment]
Thanks for the update @jovan-stojanovic !
I added a bare-bones example (examples/07_deduplication.py) to show a use case for this work, let me know if you have any questions/ideas for future directions.

**Comment by mjboos** [comment]
@jovan-stojanovic @GaelVaroquaux Initially I wanted to extend the example, but think it's better to only do that after you gave some feedback on the PR/design questions I raised above.

Seems like the test is failing due to some missing approval/workflow configuration - let me know if it's something on my end.

**Comment by GaelVaroquaux** [comment]
I approved running the tests.

**Comment by GaelVaroquaux** [comment]
Trivial comment: the file should be named "_deduplicate.py", to signal that it is private and that people should not be importing from it.

File: dirty_cat/deduplicate.py
Comment (GaelVaroquaux) [inline dirty_cat/deduplicate.py:150]: As this is used in the "deduplicate" function, convention goes that it should come before: define the functions that you need before using them. It used to be necessary for compilation, but it no longer is.

File: dirty_cat/test/test_deduplicate.py
Comment (GaelVaroquaux) [inline dirty_cat/test/test_deduplicate.py:28]: You need to use a seeded local random number generator: at the beginning of the function, define
```python
rng = np.random.RandomState(seed)
```
and use it's methods

File: examples/08_deduplication.py
Comment (GaelVaroquaux) [inline examples/08_deduplication.py:94]: I would import the function deduplicate from the main `__init__` of dirty-cat. This is where we have been putting all our user-facing functions and objects so far.

File: examples/08_deduplication.py
Comment (GaelVaroquaux) [inline examples/08_deduplication.py:115]: Can you use sphinx-gallery markup here to do titles and line separation. A bit of doc here:
https://sphinx-gallery.github.io/stable/syntax.html#embedding-rst
(it's mostly using restructured text in the comments, with the proper separator at the top)

**Comment by GaelVaroquaux** [comment]
I had a quick look, and overall things look good. I made a bunch of minor comments, but nothing on the overall design. Let me try to think about it more. Yet, so far, so good!

File: dirty_cat/deduplicate.py
Comment (GaelVaroquaux) [inline dirty_cat/deduplicate.py:144]: Don't use np.repeat as you need a list. Use multiplication on a list:
n * [1, ], this is better because you need a list at the end of the day, rather than a numpy array

**Comment by mjboos** [comment]
Thanks for the feedback @GaelVaroquaux I implemented the suggestions

File: dirty_cat/_deduplicate.py
Comment (jovan-stojanovic) [inline dirty_cat/_deduplicate.py:69]: On the docs, best is to see what was used before ([e.g. fuzzy_join](https://github.com/dirty-cat/dirty_cat/blob/master/dirty_cat/_fuzzy_join.py)) and try to implement the same. 
We try to stay consistent with work on previous features. For instance, this should be:
```suggestion
    n_clusters : typing.Optional[int], default=None
        Number of clusters to use for hierarchical clustering, if None use the
        number of clusters that lead to the lowest silhouette score.
```

File: dirty_cat/_deduplicate.py
Comment (jovan-stojanovic) [inline dirty_cat/_deduplicate.py:82]: What (object name) is returned?

File: dirty_cat/_deduplicate.py
Comment (jovan-stojanovic) [inline dirty_cat/_deduplicate.py:102]: I had the same idea for fuzzy_join, but some users pointed out that the simplest was to return one table, with an additional column containing "non-translated" values. It may be best to do here as well. From experience, the user may then more easily observe results and correct if needed.

**Review by jovan-stojanovic — CHANGES_REQUESTED** [review]
Thanks for the great work! Some comments from my side as well.
On your open question, I believe this should be a function, so I agree with what you did.

File: dirty_cat/_deduplicate.py
Comment (jovan-stojanovic) [inline dirty_cat/_deduplicate.py:16]: I think the three support functions guess_clusters, compute_ngram_distance, create_spelling_correction should be made private (add a _ before the name)

File: dirty_cat/_deduplicate.py
Comment (jovan-stojanovic) [inline dirty_cat/_deduplicate.py:40]: This is equivalent to using scikit-learn's [TDIDF Vectorizer](https://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.TfidfVectorizer.html).

File: dirty_cat/_deduplicate.py
Comment (jovan-stojanovic) [inline dirty_cat/_deduplicate.py:127]: Why did you choose criterion `maxclust`?

File: dirty_cat/_deduplicate.py
Comment (jovan-stojanovic) [inline dirty_cat/_deduplicate.py:151]: This is no longer a parameter, right?

**Review by jovan-stojanovic — COMMENTED** [review]
Some more comments.This is going to be a great new feature, thanks.

File: dirty_cat/_deduplicate.py
Comment (mjboos) [reply dirty_cat/_deduplicate.py]: Not sure I understand, you mean returning translated data as a table with suggested translations (or non-translated values)?
Like `translation_table`, just for all data instead of the unique examples (`translation_table` contains unique strings/categories + suggested translations, or the original if no translation was suggested)?

File: dirty_cat/_deduplicate.py
Comment (mjboos) [reply dirty_cat/_deduplicate.py:69]: More out of ease-of-use than performance considerations.
I found it easier for me (or other users) to "guess" the number of unique categories that is present in the data, than have an intuition about minimum cluster distance etc.

**Comment by mjboos** [comment]
Thanks for the comments @jovan-stojanovic !
I implemented your suggestions, but had some additional questions (took some time to get back to it, life happens).

When I used the function in the past month, I stumbled across another use case:
Often I notice dirty categorical data in a database/data warehouse, this data is usually too large to download and put into the function itself. So I already create a table of unique values and their counts and then use this as input (instead of the raw data).
To fix typos, I create the translation table and use it to join in the database.
I'm not sure if this is a workflow you want to support in dirty cat, but I could adapt the function to accept already unique strings + their counts (instead of the raw data/list of strings).
Trade off is that we overload the parameter (can be either unique values + counts or just list of potentially duplicate strings) and I'm not sure whether dirty cat is often used with data too large to fit into memory.

File: dirty_cat/_deduplicate.py
Comment (jovan-stojanovic) [reply dirty_cat/_deduplicate.py]: I actually meant returning a merged table `unrolled_corrections` + `translation_table` if `return_translation_table` is True.

Comment (jovan-stojanovic) [reply dirty_cat/_deduplicate.py]: Ok actually, I think the best is that this function only returns `unrolled_corrections table`. 
The main problem is that the output format changes depending on parameters value, which is usually not recommended.
Translation tables can be easily extracted then by dropping duplicates in the dirty data column..

**Comment by jovan-stojanovic** [comment]
Thanks! Interesting use case but there is something I don't understand:
> So I already create a table of unique values and their counts and then use this as input (instead of the raw data). To fix typos, I create the translation table and use it to join in the database. 

Isn't this exactly what the `deduplicate` function does? So the only difference would be that you would manually provide `unique words` and `counts` instead of doing what is in the first line `unique_words, counts = np.unique(data, return_counts=True)`
But then again, you would need to do this to get the counts and uniques, only outside of the function?

File: dirty_cat/_deduplicate.py
Comment (mjboos) [reply dirty_cat/_deduplicate.py]: I agree, it should always return the same output type.
Okay, will return the table you described above with translated/corrected and uncorrected values (lmk if I misunderstood).

**Comment by mjboos** [comment]
> Thanks! Interesting use case but there is something I don't understand:
> 
> > So I already create a table of unique values and their counts and then use this as input (instead of the raw data). To fix typos, I create the translation table and use it to join in the database.
> 
> Isn't this exactly what the `deduplicate` function does? So the only difference would be that you would manually provide `unique words` and `counts` instead of doing what is in the first line `unique_words, counts = np.unique(data, return_counts=True)` But then again, you would need to do this to get the counts and uniques, only outside of the function?

Exactly, (locally) I allow `counts` as optional argument, if it is present I assume `data` already refers to unique strings.
I get the counts & uniques via a cloud database (e.g., google cloud) and download only these. For me, this is helpful because I have very large datasets in the cloud and don't want to get the full, raw data - the correction then happens in the cloud too, by uploading the `translation_table` as a literal table and joining on it.
It's quite useful for me, but I'm not sure how many `dirty_cat` users will use a similar workflow (hence trade-off of overloading the parameters vs. supporting this workflow).

**Comment by jovan-stojanovic** [comment]
I see, then best to do in my opinion is the following:
- finish this version of deduplication;
- add an issue mentioning this, to keep track of it and see if it affect others.

**Comment by mjboos** [comment]
> I see, then best to do in my opinion is the following:
> 
> * finish this version of deduplication;
> * add an issue mentioning this, to keep track of it and see if it affect others.

Sounds good @jovan-stojanovic .
I changed the function to always return the deduplicated data and adapted the example to show how to re-create the translation table - that should make it easier to use.
Lmk if that's enough.

File: dirty_cat/_deduplicate.py
Comment (LilianBoulard) [inline dirty_cat/_deduplicate.py:41]: Just a restyle
```suggestion
    encoded = (
        TfidfVectorizer(ngram_range=ngram_range, analyzer=analyzer)
        .fit_transform(unique_words)
    )
```

File: examples/06_deduplication.py
Comment (LilianBoulard) [inline examples/06_deduplication.py:67]:
```suggestion
# set seed for reproducibility
np.random.seed(123)
```

File: examples/06_deduplication.py
Comment (LilianBoulard) [inline examples/06_deduplication.py:146]: Is there a specific reason why we need to get/store the function output? Otherwise it's much cleaner to do
```suggestion
plt.xlabel("Medication name")
plt.ylabel("Counts")
```

File: examples/06_deduplication.py
Comment (LilianBoulard) [inline examples/06_deduplication.py:90]: Same thing here
```suggestion
plt.xlabel("Medication name")
plt.ylabel("Counts")
```

File: examples/06_deduplication.py
Comment (LilianBoulard) [inline examples/06_deduplication.py:176]:
```suggestion
        print(
            f"Guessed correct spelling: {c!r} for"
            f" {spell_correct[spell_correct==c].index.values}"
        )
```

**Review by LilianBoulard — APPROVED** [review]
Excellent PR, thanks a lot!

File: examples/06_deduplication.py
Comment (mjboos) [reply examples/06_deduplication.py:147]: Just personal preference to not print the label to stdout (and have it show up in the cell).
But it's probably easier to get rid of this idiosyncrasy :)

File: examples/06_deduplication.py
Comment (mjboos) [reply examples/06_deduplication.py:177]: Not sure why `!r` is needed - `c` will be strings -> no `repr` call required?

**Comment by mjboos** [comment]
Thanks @jovan-stojanovic @LilianBoulard
I accepted all suggestions apart from the one where I had an additional question :)

File: examples/06_deduplication.py
Comment (LilianBoulard) [reply examples/06_deduplication.py:177]: Good point, the repr call is actually a way to put the value between single quotes
```
>>> print(f"Got unexpected value: {s}")
Got unexpected value: value
>>> print(f"Got unexpected value: {s!r}")
Got unexpected value: 'value'
```

Comment (mjboos) [reply examples/06_deduplication.py:177]: I see, thanks for the explanation, makes sense

File: dirty_cat/_deduplicate.py
Comment (jovan-stojanovic) [inline dirty_cat/_deduplicate.py:16]:
```suggestion
def compute_ngram_distance(
```
You need to make this function public (as it is used in the example).

File: examples/06_deduplication.py
Comment (jovan-stojanovic) [inline examples/06_deduplication.py:114]:
```suggestion
# Below we use a heatmap to visualize the pairwise-distance between medication names.
```
'We' should used instead of 'I' in examples.

File: examples/06_deduplication.py
Comment (jovan-stojanovic) [inline examples/06_deduplication.py:107]: Function to be made public (see above)
```suggestion
from dirty_cat._deduplicate import compute_ngram_distance
from scipy.spatial.distance import squareform

ngram_distances = compute_ngram_distance(unique_examples)
```

File: examples/06_deduplication.py
Comment (jovan-stojanovic) [inline examples/06_deduplication.py:64]: Best is to move this function into [datasets._fetching](https://github.com/dirty-cat/dirty_cat/blob/main/dirty_cat/datasets/_fetching.py) and rename it to something like `make_deduplication` so that we use it whenever we want to generate an example for this.
And to have a one-line import rather than a function makes the important features of this example much more visible.

Note: I see you used the same function in the [tests](https://github.com/dirty-cat/dirty_cat/blob/09a88c4053a031965ec8f08c6331491cceec5a51/dirty_cat/tests/test_deduplicate.py#L18), so maybe import it there as well

File: dirty_cat/tests/test_deduplicate.py
Comment (jovan-stojanovic) [inline dirty_cat/tests/test_deduplicate.py:40]: Import this function from datasets, see comment in example section.

**Review by jovan-stojanovic — CHANGES_REQUESTED** [review]
A few remarks came up to us just before closing this:

Hope this is it this time..

**Comment by GaelVaroquaux** [comment]
> You need to make this function public (as it is used in the example).

It implies: changing the naming, importing it in the main namespace (init) of dirty_cat and documenting it.

> Best is to move this function into datasets._fetching and rename it to something like make_deduplication

I'd rather use as a name "make_deduplication_data"

**Comment by LilianBoulard** [comment]
Could you also add a changelog major change entry for version 0.4.0? Thanks!

**Comment by LilianBoulard** [comment]
Hey @mjboos
We are looking to release a new version of dirty_cat within the next two weeks, and we're excited to have the deduplication as part of 0.4! Thus we need to merge this PR :)
Do you need any help/guidance on how to operate the previously suggested changes?
Thanks!

**Comment by mjboos** [comment]
Sure, I can prioritize it and get it done in the next few days :)

Am Mo., 9. Jan. 2023 um 14:50 Uhr schrieb Lilian ***@***.***>:

**Comment by jovan-stojanovic** [comment]
Hey @mjboos, we are ready for a beta release, would be good to have this as well. Do you need assistance with the few changes left?

File: dirty_cat/tests/test_deduplicate.py
Comment (mjboos) [reply dirty_cat/tests/test_deduplicate.py:40]: I left the function, because it's slightly different - for ease of testing I pass the randomstate explicitly here, but for the user facing function in `datasets` this makes it unncecessarily complicated IMO.
If you prefer to make it DRY-er, I can add the randomstate argument to the function in `datasets` as well.

**Comment by mjboos** [comment]
Sure, I accepted the suggestions and made the function public. I kept the different function for testing though @jovan-stojanovic and detailed why I did above (but can change it).
Also added a change log @LilianBoulard lmk if you want to have it another way or want to add attribution to suggesters of changes as well.

**Review by jovan-stojanovic — APPROVED** [review]
Thanks, will merge when the tests pass!

File: dirty_cat/tests/test_deduplicate.py
Comment (jovan-stojanovic) [reply dirty_cat/tests/test_deduplicate.py:40]: Fine for me. I think the most important was to make the example lighter. If this one is different you can keep it in the tests.

**Comment by mjboos** [comment]
> The test_docstrings.py are failing (you should update the docstrings to meet the numpydoc standards).

Oops, fixed it.

**Comment by jovan-stojanovic** [comment]
It's green, merging. Thanks!

**Pull request merged**
